### DISCUSSION TITLE
cubefs/blobstore：blobnode 在升级重启时，如何保证 ec volume 下所有的 blob 都可读？

### DISCUSSION THREAD

**Discussion by jianwei1216** [cubefs/cubefs#3403] [Q&A] [answered]
* volume EC 策略 :   8+3
* quorum write min_size ：8 + 1 = 9 ???

* blobnode
  * HDD 个数: 32 * 16 TB
  * 单个 HDD 承载的最大 chunk 文件数：1000 * 16 GB

* 必要条件：带业务在线升级 blobnode

* ec(8+3), quorum_write_min_size=9
* volume-x 在写入 blob-y 时，写成功了 9 个，给业务返回成功
* 运行一段时间后，数据还未来得及修复，又坏了一块 HDD，blob-y 剩余 8 个 ec 分片
* 随后要进行软件升级，重启 blobnode-z 上的服务进程

* 核心问题：在重启 blobnode-z 升级服务时，如何保证 blob-y 还能够 GET 读取？
  * blob-y 数据其实没有丢，但是影响了用户访问 blob-y 的可用性
* 问题1：blobnode-z 重启服务时，有办法知道所 volume-x 下所有的 blob 对象都满足 8+1 吗？
* 问题2：ec(8+3)， quorum_write_min_size 设置为 9 是否正确？

@Cloudstriff  @pengtianyue025  @sejust  @true1064  期待各位朋友，答疑解惑

**Comment by Cloudstriff** [comment, marked as answer]
1. 8+3一般用10作为put quorum，才能满足在线写入可靠性
2. 升级过程是逐台升级的，打开host_aware开关情况下，也就是最多额外容忍1台机器故障或者一台机器坏盘，因此升级服务前，最好检查下当前集群的坏盘或者故障节点情况再做升级。

PS：可以考虑增加校验块的个数来提升故障、坏盘等情况下的系统在线可用性

**Reply by jianwei1216** [reply]
> 8+3一般用10作为put quorum，才能满足在线写入可靠性

在选择 volume 时，一旦由于网络延迟或者其他问题，volume 无法满足 8+2=10 的要求，这就要求 access 必须重试用户的写请求?

**Reply by sejust** [reply]
> > 8+3一般用10作为put quorum，才能满足在线写入可靠性
> 
> 在选择 volume 时，一旦由于网络延迟或者其他问题，volume 无法满足 8+2=10 的要求，这就要求 access 必须重试用户的写请求?

access 会根据写入错误，把这个 volume 暂停写一段时间，或剔除

**Reply by jianwei1216** [reply]
用户本次的写请求还会重新选择poolset写入吗？

**Reply by sejust** [reply]
> 用户本次的写请求还会重新选择poolset写入吗？

会有几次重试，重试失败后就返回给用户

**Reply by jianwei1216** [reply]
即使 write_quorum=10，在重启blobnode-z升级软件时，当前的实现其实也无法知道volume-x下所有的blob都满足8+1，

当数据还没来得及修复时，
blob-y可能只有8个分片，而非8+1个分片，

在blobnode-z重启期间，因为剩余8-1个分片在线，blob-y无法读取出来


解决方案：

计划升级重启blobnode-z前，先扫出来该节点上有哪些volume（流程控制），

针对这些volume下所有blob进行k+m个分片扫描比对（及时扫描机制），

当blob被扫描出不满足k+1或k+2时，

在重启升级软件前，优先恢复这些丢失分片数多的blob（及时恢复机制），

等它们修复完成后，再扫描比对volume下的blob，都满足k+1（重启的前置条件），重启升级软件。

**Reply by sejust** [reply]
当前没有在程序中实现手动检查的工具；重启过程也是人为的。
这个问题可以在运营过程中尽量保证：
1、所有blob会有定期巡检；有可能出现坏块，但不会持续很久且单个volume基本不会出现同时两个以上坏的分区
2、逐台重启也是可以人为控制的
3、不会在系统中有很多修复任务、坏盘、下线过程中作过度重启

**Reply by jianwei1216** [reply]
机械硬盘的性能有限，（200MBps 200IOPS基本上是理想，经过各种软件栈未必能够达到，如果再加上长期的写删，磁盘碎片化在所难免，最重要的是影响读的时延）





在单盘高水位，16TB， 比如80%场景下，

在线ec会将文件切的更碎，基本上单盘就是亿级别的blob

修复时的随机读应该会比较严重，进而影响数据修复效率，


在高水位下，当前实现坏一块机械盘多长时间能够恢复？

1小时
5小时
10小时
12小时
24小时？


集群上线3年后，机械硬盘的老化速度应该会更高，

第一块坏盘还未彻底修好？

第二块坏盘又出现了


全盘修复时间T决定了同时遇到2个坏盘的概率还是有的


讨论的目的:
排除磁盘静默错误

**（一）针对quorum系统，是一种无状态的系统（要想知道状态只能靠扫描比对一遍才知道缺失分片的分布）
我们在运维阶段如何能够快速的确切的知道被运维节点的实时状态(应该只能靠扫描比对)，
人能不能执行运维操作，确切的知道执行运维操作会有什么后果？
运维操作导致用户不可读，应该是最差的结果了。**

**k+m，我觉得至少满足k+2，才能够极大概率保证运维过程中的级联故障不会影响业务的可读性，k+1都不一定能够避免级联故障。
这也是为什么我们把 ec(8+3) 的 quorum_write 设置为 8+2 的原因。**


(二）针对 raft/ceph(mon+osdmap+crush+pg peering) ，是一种有状态的系统（它在写入过程中维护了 raftlog/pglog，它想知道状态就很方便，很快的比较 log  的差异和提交的差异就可以了）
我们在运维过程中，比如 ceph -s 一下就可以知道 pg 的状态了，就能够知道这个节点能不能重启了
所以它的运维操作是确定性的


对于对象存储，quorum 系统无疑是占据天时地利人和的，

这是 2 类不同的设计理念，
（一）quorum系统 IO路径放弃了强一致性写入，
   - 收益：低尾延迟，更高的 ec(k+m) 比例，跨AZ部署更稳定；
   - 代价：要想知道集群的状态，必须得扫描比对一遍，扫描比对效率决定了数据的安全性

（二）类raft/ceph系统 在IO路径中，保证了强一致性写入，维护了 raftlog/pglog，确保能够通过 log 选出权威，副本分片与权威比较即可
   -  收益：副本/ec分片间的强一致，高效的维护集群状态
   - 代价：高尾延迟，较低的 ec(k+m) 比例，跨 AZ 部署不一定稳定，性能受尾延迟影响；

**Reply by jianwei1216** [reply]
> 即使 write_quorum=10，在重启blobnode-z升级软件时，当前的实现其实也无法知道volume-x下所有的blob都满足8+1，
> 
> 当数据还没来得及修复时， blob-y可能只有8个分片，而非8+1个分片，
> 
> 在blobnode-z重启期间，因为剩余8-1个分片在线，blob-y无法读取出来
> 
> 解决方案：
> 
> 计划升级重启blobnode-z前，先扫出来该节点上有哪些volume（流程控制），
> 
> 针对这些volume下所有blob进行k+m个分片扫描比对（及时扫描机制），
> 
> 当blob被扫描出不满足k+1或k+2时，
> 
> 在重启升级软件前，优先恢复这些丢失分片数多的blob（及时恢复机制），
> 
> 等它们修复完成后，再扫描比对volume下的blob，都满足k+1（重启的前置条件），重启升级软件。

**针对 quorum 系统，在运维阶段，开发出及时了解当前节点状态的扫描和修复工具，你们觉得有必要吗？**

**Reply by JasonHu520** [reply]
现在blobstore本身支持数据和元数据的扫描，都是作为后台任务在服务不繁忙的时候可以打开，scheduler这个模块会扫描元数据是否有缺失，发现缺失会及时修补；至于数据层面blobnode模块也有相应的后台巡检任务，发现坏盘或者数据损坏也会及时修复

**Reply by Cloudstriff** [reply]
没有太大必要，如果很担心硬盘损坏导致在线可用性和数据可靠性影响，应该在编码策略上调整，采用扇出更大的模式，比如12+4、12+5等，在保证较低数据冗余度（成本）的情况下，容忍更多的节点硬盘损坏和升级重启影响。

**Reply by sejust** [reply]
> m 系统，在运维阶段，开发出及时了解当前节点状态的扫描和修复工具，你们觉得有必要吗？

如果确有顾虑是可以实现的，只需要再 cli 下加一点命名行工具就能实现，
比如想检查某台机器所有volume的健康状态：

1. 从 clustermgr 相关接口列举出当前机器所有磁盘
2. 再 clustermgr 列举每一块磁盘上所有的卷，及扇出 ec 对应的数据位置 vuid
3. 再获取每一个vuid 的状态

以上都是有现有接口，但一次操作扇出很大，代价也是比较高的。
如 8 + 3 模式，每个 chunk 16G， 18T * 24 盘：
18000 / 16  * 24  * 11  至少要发送 30W+ 次 chunk stat 请求

运营人员重启一台机器就导致读不可用，说明当前整个系统可用性已经非常低。
另外，线上系统修复 12T 数据盘，大约需要3-4小时。

**Reply by jianwei1216** [reply]
扫描一块 16 TB 的盘，16GB chunk

单盘：16000 / 16 = 1000 个 chunk

EC：8+3

bid 在 blobstore 中应该是可以全内存的，理论上应该会很快，我看 clustermgr 在 inspector 扫描比对时，task 只等待 10 秒钟

16TB的磁盘，扫描比对 (8+3) * 1000 个 volume 需要多久？

单盘所有 volume 扫描的效率直接决定了这个事情的可行性。

**Reply by jianwei1216** [reply]
> 运营人员重启一台机器就导致读不可用，说明当前整个系统可用性已经非常低。

及时扫描比对查看系统状态的工具的必要性还是有的，上线时间较长3～5年后的系统，坏盘率高，这类工具就是运维的依赖了。

这个工具可以是一个补偿性的检查
